content/posts/0006-java-rearchitecture.md

```diff
@@ -0,0 +1,194 @@
+
+`BPL_JAVA_NMT_LEVEL` - set to `summary` by default, this specifies the level of detail for the memory usage data captured. For a more detailed output, it can be set to `detail`.
+
+The following default NMT arguments are passed to the JVM:
+
+```
+-XX:+UnlockDiagnosticVMOptions
+-XX:NativeMemoryTracking=summary
+-XX:+PrintNMTStatistics`
+```
+
+Example:
+
+To start a new bash session and capture NMT data from a running container run the following:
+
+`docker exec -it <container-id> /cnb/lifecycle/launcher /bin/bash`
+
+If a JDK was installed, this session should have the JDK tool `jcmd` on the $PATH variable. To capture NMT data you can run:
+
+`jcmd 1 VM.native_memory summary`
+
+NMT data will also be printed at JVM exit:
+
+![NMT](/images/posts/0006/nmt.png)
+
+#### **Java Flight Recorder (JFR)**
+
+Java Flight Recorder is a tool which is integrated into the JVM and allows for the collection of profiling and diagnostic data about your application. Support for this feature has been added, however, it is disabled by default. To enable it, simply use the environment variable:
+
+`BPL_JFR_ENABLED` - set to `false` by default
+ 
+The tool supports a range of configuration options which can be set using the following environment variable:
+
+`BPL_JFR_ARGS` - this supports a comma-separated list of options (full list detailed in the Oracle docs [here](https://docs.oracle.com/javacomponents/jmc-5-4/jfr-runtime-guide/comline.htm#JFRUH193)). If the tool is enabled but no options are specified here, the buildpack will at a minimum configure the tool to write data to a temporary file on JVM exit. The following default options will be configured to allow this:
+
+`dumponexit=true` - this tells the JFR tool to write recording data to a file when the JVM exits.
+`filename=<system-temp-dir>/recording.jfr` - the location of the dump file, on linux this will be “/tmp/recording.jfr”
+
+If any custom arguments are specified, the above defaults are not applied. When enabled, a JFR argument will be passed to the JVM, for example, the defaults result in the following:
+
+`-XX:StartFlightRecording=dumponexit=true,filename=/tmp/recording.jfr`
+
+Example:
+
+To start a docker container with JFR enabled, setting some custom arguments, run the following:
+
+```
+docker run \
+--env BPL_JFR_ENABLED=true \
+--env BPL_JFR_ARGS=filename=/tmp/my-recording.jfr,duration=60s \
+samples/java
+```
+
+When you wish to retrieve the recording file, you can run the following to copy it from the container (the location specified in the above 'filename' argument) to your local current directory:
+
+```
+docker cp <container-id>:/tmp/my-recording.jfr .
+```
+
+### Newly Integrated Features
+
+In addition to the new features now supported by the buildpack, two features which previously a user to rebuild their image are now integrated into the Paketo Java Buildpack directly. These can simply be enabled at runtime using an environment variable.
```

Comment:
Small typo in here:
"two features which previously a user to rebuild their image are now..."